### PR TITLE
ptstorage: change Protect and GetRecord to work with target column

### PR DESCRIPTION
**Pull request by adityamaru**
Protected timestamp records now apply on a `ptpb.Target` instead
of `roachpb.Spans`. This change modifies the `Protect` and `GetRecord`
methods of the ptstorage interface. Protect will persist the
`ptpb.Target` specified on the passed in `ptpb.Record` as an encoded
protocol buffer in the `target` column, and `GetRecord` will read the
`target` column when constructing a `ptpb.Record` to return to the caller.

If we are running in a mixed version state where the migration that adds
the `target` column to the system.pts_records table has not run,
we use the existing logic to protect Spans. This logic has wholesale
been moved to `deprecatedProtect` method.

Once the migration has run, all calls to protect must
have a non-nil `ptpb.Target` field on the record, as we will no longer
be persisting the `DeprecatedSpans` field in the underlying system table.
Post migration, we do not use the `kv.protectedts.max_spans` to enforce
any limits, but continue to use `kv.protectedts.max_bytes` to enforce a
ceiling on the total size of encoded targets we can store in the subsystem.

Most tests in `ptstorage_test` continue to run in both a pre and post migration
state to ensure that `Storage` is handling both span backed and target backed
records correctly. This can be cleaned up in a future version when we get
rid of spans from the pts record entirely. All other PTS tests in the `kvserver`
package run on a version prior to the pts migration since the remainder of the
new pts system is yet to be implemented. These tests will be revisited as and
when we develop those pieces of the subsystem.

Informs: #73727

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
First three commits from #74248.

**Comment by adityamaru** [comment]
First commit is from https://github.com/cockroachdb/cockroach/pull/74248.

**Comment by adityamaru** [comment]
@ajwerner @arulajmani friendly ping for a review on this!

**Review by ajwerner — COMMENTED** [review]
cursory pass this looks fine 

> Reviewed 6 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-:-MtJ0V3tChYnQo4GXy09:b-3u6su3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, @irfansharif, and @msbutler)

___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtJ0-368ukFupk_D7ee:-MtJ0-368ukFupk_D7ef:b-odpfyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L146)):*
> ```Go
> 	}
> 
> 	// TODO(during review): To avoid breaking protected timestamps before the
> ```

How hard is it to put in a cluster setting or testing knob that we pair with the version gate in all the places we check it that we default to false? With that, we wouldn't break anything and we can test the new logic by enabling that cluster setting. Then, before we release, we get rid of the cluster setting or testing knob.

**Review by arulajmani — COMMENTED** [review]
> Reviewed 6 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74297)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, and @msbutler)

___
*[pkg/kv/kvserver/protectedts/ptstorage/sql.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtIwY820lfXSyAT9zIa:-MtIwY820lfXSyAT9zIb:b-3jnhe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/sql.go#L126)):*
> ```Go
> `
> 
> 	getRecordsWithoutTargetQueryBase = `
> ```

nit: might be worth commenting that the `Target` column was added in a migration which necessitates these with and without target versions of queries.
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtIweVl9AWnYhNJIiTl:-MtIweVmCw1d_stGmumR:b5n20lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L83)):*
> ```Go
> 		return errors.Wrap(err, "failed to marshal spans")
> 	}
> 	it, err := p.ex.QueryIteratorEx(ctx, "protectedts-protect", txn,
> ```

nit: s/protectedts-protected/protectedts-deprecated-protect/
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtIx3OX4xb0xC7x8SY_:-MtIx3OX4xb0xC7x8SYa:b-cqjrm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L142)):*
> ```Go
> 		meta = []byte{}
> 	}
> 	if !p.settings.Version.IsActive(ctx, clusterversion.AlterSystemProtectedTimestampAddColumn) {
> ```

Might be worth adding a comment about this here as well and a TODO to remove this in v22.2.
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtJ0-368ukFupk_D7ee:-MtJ3vYg43FjZrinrVl8:bfizgdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L146)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

How hard is it to put in a cluster setting or testing knob that we pair with the version gate in all the places we check it that we default to false? With that, we wouldn't break anything and we can test the new logic by enabling that cluster setting. Then, before we release, we get rid of the cluster setting or testing knob. 
</blockquote></details>

We could also run some of the tests that are supposed to work before and after the`AlterSystemProtectedTimestampAddColumn` migration with the testing knob set to both values for 22.2. 
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtIxKYnC4uLSYSBzEpN:-MtIxKYnC4uLSYSBzEpO:biimq9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L213)):*
> ```Go
> 		}
> 	} else {
> 		row, err = p.ex.QueryRowEx(ctx, "protectedts-GetRecord", txn,
> ```

nit: Similar to how you did for `Protect`, you could pull this logic out in a deprecated method as well. That way, when we come back to this to cleanup, it'll be easier.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74297)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @irfansharif, and @msbutler)

___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtIx3OX4xb0xC7x8SY_:-MtJYp3a9vnGq7VMvlhk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L142)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Might be worth adding a comment about this here as well and a TODO to remove this in v22.2.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtJ0-368ukFupk_D7ee:-MtJgbLi49D__cF6Gq_K:bkljlfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f16f9e367deb1d7094bec5844352218dd460bec/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L146)):*
> Done.
Adding a testing knob that currently defaults to false. All tests that want to run with the new system have to override this knob. Prior to the release, we will default this knob to true or rename it to `DisablePTSForMultiTenant` and have all the tests that want to run with the old system override it.

**Comment by adityamaru** [comment]
I'm not sure what happened to reviewable, but I have addressed all the comments so this if RFAL.

File: pkg/kv/kvserver/protectedts/ptstorage/storage.go [hunk 1]
```diff
@@ -225,18 +324,47 @@ func (p *storage) GetState(ctx context.Context, txn *kv.Txn) (ptpb.State, error)
...
 func (p *storage) getRecords(ctx context.Context, txn *kv.Txn) ([]ptpb.Record, error) {
+	if !p.settings.Version.IsActive(ctx, clusterversion.AlterSystemProtectedTimestampAddColumn) ||
+		!p.knobs.EnableProtectedTimestampForMultiTenant {
+		return p.deprecatedGetRecords(ctx, txn)
+	}
+
 	it, err := p.ex.QueryIteratorEx(ctx, "protectedts-GetRecords", txn,
 		sessiondata.InternalExecutorOverride{User: security.NodeUserName()},
-		getRecordsQuery)
+		getRecordsWithoutTargetQuery)
```
Comment (adityamaru) [inline pkg/kv/kvserver/protectedts/ptstorage/storage.go:358]: this is wrong, it should be getRecordsQuery. I'm just waiting for CI to shake out all failures before repushing.

Comment (adityamaru) [reply pkg/kv/kvserver/protectedts/ptstorage/storage.go]: done.

**Review by arulajmani — APPROVED** [review]
Couple of minor comments, but otherwise :lgtm: 

> Reviewed 1 of 19 files at r1, 15 of 24 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-:-MtOL04T5WAtnrg8-qu6:b-w4b3zs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @arulajmani, @irfansharif, and @msbutler)

___
*[pkg/kv/kvserver/protectedts/ptprovider/provider.go, line 47 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOJVIxB7LMyJBnV5DN:-MtOJVIxB7LMyJBnV5DO:baa766s) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/kv/kvserver/protectedts/ptprovider/provider.go#L47)):*
> ```Go
> 
> // New creates a new protectedts.Provider.
> func New(cfg Config, knobs *protectedts.TestingKnobs) (protectedts.Provider, error) {
> ```

Any reason to not put these on the `Config` struct itself? 
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOIP80AqlQUGfS4-h6:-MtOIP80AqlQUGfS4-h7:b-9vzyy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L156)):*
> ```Go
> 	// TODO(adityamaru): Delete in 22.2 once we exclusively protect `target`s.
> 	if !p.settings.Version.IsActive(ctx, clusterversion.AlterSystemProtectedTimestampAddColumn) ||
> 		!p.knobs.EnableProtectedTimestampForMultiTenant {
> ```

nit: maybe pull this logic into a `useDeprecatedProtectedTSStorage` or something -- we use it in a couple of places and it's slightly more readable than this disjunction. 
___
*[pkg/ccl/backupccl/backup_test.go, line 6759 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOJBrn14FgW30NTgnV:-MtOJBrn14FgW30NTgnW:b-hay66i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/ccl/backupccl/backup_test.go#L6759)):*
> ```Go
> 	dir, dirCleanupFn := testutils.TempDir(t)
> 	defer dirCleanupFn()
> 	clusterArgs := base.TestClusterArgs{}
> ```

revert?
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage_test.go, line 324 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOKedd5s0GT2ENFsLA:-MtOKedd5s0GT2ENFsLB:buuhw7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/kv/kvserver/protectedts/ptstorage/storage_test.go#L324)):*
> ```Go
> 
> 	// If set to true, the test will be run on a cluster version prior to
> 	// AlterSystemProtectedTimestampAddColumn, thereby testing the "old" protected
> ```

Maybe reference the testing knob instead?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74297)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @arulajmani, @irfansharif, and @msbutler)

___
*[pkg/kv/kvserver/protectedts/ptprovider/provider.go, line 47 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOJVIxB7LMyJBnV5DN:-MtP_Q8jC8AKYFL4Ipyo:bl7tnra) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/kv/kvserver/protectedts/ptprovider/provider.go#L47)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Any reason to not put these on the `Config` struct itself? 
</blockquote></details>

good point, done.
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage.go, line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOIP80AqlQUGfS4-h6:-MtPcz-u8cHiy_vg_ygQ:b76ehqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L156)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: maybe pull this logic into a `useDeprecatedProtectedTSStorage` or something -- we use it in a couple of places and it's slightly more readable than this disjunction. 
</blockquote></details>

i like it, thanks!
___
*[pkg/ccl/backupccl/backup_test.go, line 6759 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOJBrn14FgW30NTgnV:-MtP_n2gBuT68fLSW912:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/ccl/backupccl/backup_test.go#L6759)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

revert?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/protectedts/ptstorage/storage_test.go, line 324 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74297#-MtOKedd5s0GT2ENFsLA:-MtPa9lVC07fNtHeYr0R:b-77ocun) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f763873f1b44ff2e1c267d0c2a897b9b6082608/pkg/kv/kvserver/protectedts/ptstorage/storage_test.go#L324)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Maybe reference the testing knob instead?
</blockquote></details>

yup stale comment, done.

**Comment by adityamaru** [comment]
@ajwerner let me know if you'd like to take another pass at this, thanks!

**Review by arulajmani — APPROVED** [review]
> Reviewed 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74297)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @arulajmani, @irfansharif, and @msbutler)

**Comment by adityamaru** [comment]
borsing to cleanup some of the PRs on top of this will keep a lookout for additional comments

TFTR!

bors r=arulajmani

**Pull request merged**